internal/overlord/logstate/loki.go

```diff
@@ -0,0 +1,250 @@
+//	if len(f.buffer.messages) > 0 {
+//		f.flushBuffer()
+//	}
+//}
+
+func (f *lokiForwarder) Stop() {
+	f.stopCh <- struct{}{}
+}
+
+func (f *lokiForwarder) flushBuffer() {
+	err := f.client.Send(f.buffer.Request())
+	if err != nil {
+		logger.Noticef(
+			"Forwarding logs to target %q: cannot send logs for service %q: %v",
+			f.targetName, f.serviceName, err)
+		// TODO: retry?
+	}
+
+	f.buffer.Reset()
+}
+
+type lokiBuffer struct {
+	labels map[string]string
+
+	buffer  *bytes.Buffer
+	encoder *json.Encoder
+
+	messages    int
+	maxMessages int
+	deadline    *time.Time
+	delay       time.Duration
+}
+
+func newLokiBuffer(labels map[string]string, maxMessages int, delay time.Duration) *lokiBuffer {
+	buf := bytes.NewBuffer(make([]byte, 0, 4096))
+
+	b := &lokiBuffer{
+		labels:      labels,
+		buffer:      buf,
+		encoder:     json.NewEncoder(buf),
+		maxMessages: maxMessages,
+		delay:       delay,
+	}
+	b.Reset()
+	return b
+}
+
+func (b *lokiBuffer) Reset() {
+	b.messages = 0
+	b.deadline = nil
+	b.buffer.Reset()
+
+	// Write start of request
+	b.buffer.WriteString(`{"streams":[{"stream":`)
+	b.encoder.Encode(b.labels)
+	b.buffer.WriteString(`,"values":[`)
+}
+
+func (b *lokiBuffer) WriteEntry(entry servicelog.Entry) {
+	// Write in format
+	//    ["<unix epoch in nanoseconds>","<log line>"]
+	b.buffer.WriteString("[")
+	timestamp := strconv.FormatInt(entry.Time.UnixNano(), 10)
+	b.encoder.Encode(timestamp)
+	b.buffer.WriteString(",")
+	message := strings.Trim(entry.Message, "\n")
+	b.encoder.Encode(message)
+	b.buffer.WriteString("],")
+
+	b.messages++
+	if b.deadline == nil {
+		// Set write deadline
+		deadline := time.Now().Add(b.delay)
+		b.deadline = &deadline
+	}
+}
+
+func (b *lokiBuffer) ReadyToFlush() bool {
+	return (b.messages >= b.maxMessages) ||
+		(b.deadline != nil && time.Now().After(*b.deadline))
+	// TODO: flush when buffer length is past a certain threshold?
+}
+
+// Prepares a JSON request for sending to Loki.
+func (b *lokiBuffer) Request() io.Reader {
+	// Need to remove last ',' in buffer
+	b.buffer.Truncate(b.buffer.Len() - 1)
+	// Write end of request
+	b.buffer.WriteString("]}]}")
+
+	return b.buffer
+}
+
+const (
+	lokiPushAPI = "/loki/api/v1/push"
+)
+
+// lokiClient handles sending requests to Loki.
+type lokiClient struct {
+	client *http.Client
+	url    string
+}
+
+func newLokiClient(location string) *lokiClient {
+	return &lokiClient{
+		client: &http.Client{
+			Timeout: 10 * time.Second,
+		},
+		url: location,
+	}
+}
+
+func (c *lokiClient) Send(body io.Reader) error {
+	resp, err := c.client.Post(c.url+lokiPushAPI, "application/json", body)
+	if err != nil {
+		// TODO: annotate this error with some more context?
+		return err
+	}
+	defer resp.Body.Close()
+
+	// Check response status code to see if it was successful
+	if c := resp.StatusCode; c >= 200 && c < 300 {
+		return nil
+	}
+
+	// Request to Loki failed
+	// TODO: retry?
+	b := make([]byte, 1024)
+	n, err := resp.Body.Read(b)
```

Comment:
We need to handle n better here and error. Error could also be EOF and their may be more data available other then what n bytes were read.

We should extract all the data out and handle all the cases of error and n properly.

See https://pkg.go.dev/io#Reader

Reply:
fair point about EOF, I'm now using an io.LimitReader instead. We don't want to extract all the data cause we don't know how large it might be. Limit it to 1KB.